addons/account/migrations/13.0.1.1/post-migration.py

```diff
@@ -248,14 +248,62 @@ def migration_invoice_moves(env):
         RETURNING aml.id""",
     )
     aml_ids = tuple(x[0] for x in env.cr.fetchall())
+    # Fix (exceptional) cases where two or more move lines points to the same old invoice line
+    while True:
+        openupgrade.logged_query(
+            env.cr, """
+            UPDATE account_move_line aml
+            SET old_invoice_line_id = NULL
+            FROM (SELECT id, rank() OVER (PARTITION BY old_invoice_line_id) repeated_invoice_line
+                  FROM account_move_line
+                  WHERE old_invoice_line_id IS NOT NULL) aml2
+            WHERE aml.id = aml2.id AND aml2.repeated_invoice_line > 1
+            RETURNING aml.id""",
+        )
+        wrong_aml_ids = tuple(x[0] for x in env.cr.fetchall())
+        if not wrong_aml_ids:
+            break
+        aml_ids = tuple(x for x in aml_ids if x not in wrong_aml_ids)
+        openupgrade.logged_query(
+            env.cr, query + """
+            LEFT JOIN account_move_line aml2 ON aml2.old_invoice_line_id = ail.id
+            """ + minimal_where + """
+                AND ail.account_id = aml.account_id
+                AND ai.commercial_partner_id = aml.partner_id
+                AND ((ail.account_analytic_id IS NULL AND aml.analytic_account_id IS NULL)
+                    OR ail.account_analytic_id = aml.analytic_account_id)
+                AND aml.old_invoice_line_id IS NULL AND aml2.id IS NULL
+            RETURNING aml.id""",
+        )
+        aml_ids += tuple(x[0] for x in env.cr.fetchall())
     # Try now with a more relaxed criteria, as it's possible that users change some data on amls
-    openupgrade.logged_query(
-        env.cr, query + minimal_where + """
-            AND aml.old_invoice_line_id IS NULL
-            AND rc.anglo_saxon_accounting IS DISTINCT FROM TRUE
-        RETURNING aml.id""",
-    )
-    aml_ids += tuple(x[0] for x in env.cr.fetchall())
+    while True:
```

Comment:
I don't think we should include 2 loops, but all in the same one

Reply:
Why? The first one is for stricter criteria and the second one is for lesser criteria. If we mix both, it could lead to undesired (although remote) cases. Performance is unaffected by doing separated or mixed way.

Reply:
There's no risk of repeat lines if you also include the join criteria on the second main SQL, but we don't repeat some code and don't do even less readable than it is. Please try to put only one loop.

Reply:
Now there is only one loop.